manage/upgrading/version_specific_migration/upgrade_to_python3.rst

```diff
@@ -0,0 +1,248 @@
+       26: q        BINPUT     2
+       28: C        SHORT_BINBYTES b'Umlaut'
+       36: q        BINPUT     3
+       38: X        BINUNICODE 'unicode'
+       50: q        BINPUT     4
+       52: X        BINUNICODE 'Ümläut'
+       65: q        BINPUT     5
+       67: u        SETITEMS   (MARK at 5)
+       68: .    STOP
+    highest protocol among opcodes = 3
+
+
+When reading a pickle created with python2 with python3 that contains non-ascii
+characters in a field declared with OPTCODE `STRING` python3 is trying to interpret it as python3 string (py2 unicode)
+and we might end up getting a UnicodeDecodeError for this pickle in ZODB.serialize
+
+
+.. code-block:: bash
+
+    $ python3
+    >>> b'\xc3\x9cml\xc3\xa4ut'.decode('ascii')
+    Traceback (most recent call last):
+      File "<stdin>", line 1, in <module>
+    UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 0: ordinal not in range(128)
+
+
+Or when utf-8 encoded byte-strings are interpreted as unicode we do not get an error but mangled non-ascii characters
+
+.. code-block:: bash
+
+    $ python3
+    >>> print('\xdcml\xe4ut')
+    Ümläut
+    >>> print('\xc3\x9cml\xc3\xa4ut')
+    ÃmlÃ¤ut
+
+
+
+
+Custom Content Types
+--------------------
+
+TODO: Not yet sure if custom types need to provide additional mappings for zodbupdate.
+
+Here is an example Pull Request that adds them: `https://github.com/zopefoundation/Products.PythonScripts/pull/19 <https://github.com/zopefoundation/Products.PythonScripts/pull/19>`_
+
+workflow: analyze, read sourcecode, add pdb to see which values are passed to attribute to decide whether to use bytes or utf-8
+
+.. code-block:: bash
+
+    bin/zodb-py3migrate-analyze py2/var/filestorage/Data.fs -b py2/var/blobstorage -v
+
+
+
+Migrate Database using zodbupdate
+---------------------------------
+
+add zodbupdate to buildout eggs::
+
+    [zodbupdate]
+    recipe = zc.recipe.egg
+    eggs =
+        ${buildout:eggs}
+        zodbupdate
+        zodb.py3migrate
+
+    scripts =
+        zodb-py3migrate-analyze
+        zodbupdate
+
+
+
+Migrate database so it can be read using Python 3.
+
+.. warning:: 
+
+    This migrates our database in place. Make sure to make a backup before!
+
+.. code-block:: bash
+
+    cd $BUILDOUT
+    bin/instance stop
+    cp var/filestorage/Data.fs var/filestorage/Data.fs-back
+    bin/zodbupdate --pack --convert-py3 --file var/filestorage/Data.fs
+
+
+
+Downtime
+''''''''
+
+This step actually requires to take your site offline or into read-only mode.
+
+
+Some thoughts on doing upgrades w/o downtime that came up in a hangout during a coding sprint in October 2018:
+
+
+- jim mentions downtime. would try to leverage the zrs replication protocol, secondary server with converted data.
+  It would probably be a trivial change to zrs.
+- for relstorage jim mentions a zrs equivalent for relstorage: http://www.newtdb.org/en/latest/topics/following.html
+- david thought out loud about
```

Comment:
@frisi you still remember what David was thinking loud about? Otherwise this sentence makes no sense.

Reply:
sorry, copy paste error. i completed it